### ISSUE TITLE
Semantic segmentation is very slow

### ISSUE THREAD

**Issue by harpone** [pytorch/xla#2199]
## 🐛 Bug

I'm training a model that's about same size as resnet-50, but doing Open Images semantic segmentation. Inputs are 256x256 pixel images, and ground truth segmentation masks are lower resolution, e.g. 32x32. 

Network is basically a resnet, except fully convolutional outputs of shape `[batch_size, num_classes, 32, 32]` etc.

Training is about 5x to 10x times slower than ImageNet. I'd expect about same speed...

Notable differences to ImageNet classification model:

- I'm using `nn.ModuleList` in a couple of places, because I'm actually outputting at different resolutions, Feature Pyramid Network style
- sometimes outputting lists of tensors, although this should not be an issue
- also doing stuff like this:

```
    def forward(self, images):
        latents = self.backbone(images)  # list of shapes (B, args.hash_dim, H / k, H / k) for k in e.g. [8, 16, 32]
        logits = [head(latent) for head, latent in zip(self.heads, latents)]

        return dict(latents=latents, logits=logits)
```

`self.backbone` is a fairly standard resnet, which works fast with ImageNet. dicts, lists etc shouldn't be an issue right?

- loss is basically just a classification loss for each of the 32*32 feature pixels, and can be implemented with `F.cross_entropy`
- using plenty of dataloader workers, so should not be capped by reading images. Also, TPU idle time varies quite a lot from 0% to 50% (closer to 0% though), when for ImageNet I'm getting consistent 10%-20% idle time with similar parameters

## To Reproduce

1. Train a semantic segmentation model, e.g. FPN-like
2. Observe weird slowness

Unfortunately I can't release the code just yet... but maybe in the near future.

Here's the link to metrics report (would be really nice to have an explanation for the various metrics somewhere btw...): https://gist.github.com/harpone/5a49666842ec917460627f650565cf40

- At least `DeviceLockWait` seems to be about 5x higher than for ImageNet 
- Also `InboundData` is 7.84GB when for ImageNet it was 2.2KB! What's that about?
- `IrValueTensorToXlaData` is also >25x higher
- `TensorsGraphSize` is a lot lower

Not sure how well these can be compared, since the ImageNet model ran for about 8x more iterations...

## Environment

 - Reproducible on XLA backend: TPU, v3-8 on a VM conda environment
 - torch_xla version: 1.5
 - debug report here: https://www.dropbox.com/s/4h711unw49w21a5/debug_report.tar?dl=0

**Comment by harpone** [comment]
OK looking at the debug data, and for the semantic segmentation model, this is what TensorsGraphSize is looking like:

![Screenshot from 2020-06-11 14-03-48](https://user-images.githubusercontent.com/5112840/84378188-6dbdac80-abec-11ea-8bb7-cbbe96dc626e.png)

For imagenet classification, that value stays constant, so seems like something funny is going on...

**Comment by harpone** [comment]
FYI just tried this with latest nightly, and the same thing happens.

**Comment by dlibenzi** [comment]
You have 58 steps, and 170 execute.
Also 124 `aten::_local_scalar_dense` for 58 steps.
I think somewhere there is an `item()` conversion.
Ideally you should have N execute for N steps, or close to.

**Comment by harpone** [comment]
Aha OK, I'll try to find the problem spot! It's really weird, because I've been modifying the code to be closer and closer to the imagenet classification (which works fast), and there's really not much difference to that anymore...

**Comment by dlibenzi** [comment]
With https://github.com/pytorch/xla/pull/2209 it will hopefully be easier to track from where calls are coming from.
